new fetch-based api-wrapper for client-side http requests

gka opened the pull request:
I was unhappy with the way the current methods in `fetch.js` are being written and started to work on a new wrapper. The new function is very opinionated and designed to work out of the box with our API. 

The differences to the old `getJSON`/`putJSON` etc methods are:

* only one function instead of several for different http methods. the method is passed as first argument
* separates the url path from the baseUrl (=domain). baseUrl defaults to our api domain for maximum convenience
* accepts both strings and objects as payload. anything that's not a string will be JSON.stringified.
* parses http response based on the response `content-type` header.
* `mode:' cors'` & `credentials: 'include'` is always set (not sure if we ever want to not authenticate)
* request content-type defaults to `application/json` but can be set via parameter 

**Usage examples**

the method is written so it can be used as one-liner in most cases

```js
// simple GET request
const team = await httpReq('/v3/team/abc');

// send DELETE request with no body
await httpReq.delete('/v3/charts/12345');
// you can also pass method with options
await httpReq('/v3/charts/12345', { method: 'delete' });

// create a new chart
const chart = await httpReq.post('/v3/charts', {
    payload: {
        title: 'My new chart',
        type: 'd3-scatterplot'
    }
});

// trigger PNG export 
const blob = await httpReq('/v3/charts/12345/export/png', { method: 'post' });
downloadjs(blob, `filename.png`, 'image/png');

// update chart data
await httpReq('/v3/chart/12345/data', {
    method: 'put',
    payload: csvData,
    headers: {
        'Content-Type': 'text/csv'
    }
});

// non-datawrapper api request
const res = await httpRequest.patch('/patch', {
    baseUrl: 'https://httpbin.org',
    payload: { answer: 42 }
});
```

gka commented:
refactored the method based on some preliminary feedback from @sto3psl over lunch

* the `options` are now matching the fetch options and are passed directly to the `fetch` call, giving the caller complete control over the config while still providing sensible defaults
* the only non-fetch options mixed in are
  1. `raw` (which bypasses the automatic response parsing if set to `true`)
  2. `baseUrl` (which is prepended to the `path` and defaults to our api domain)
  3. `payload` (which can be used to pass object that will JSON.stringified as `body`)
* instead of guessing whether or not to json-stringify the payload the caller can now simply pass a `body` instead of `payload`

Perhaps most controversially I added support for two different argument patterns, one being (preferred by me)

```js
httpReq(method, path, options)
```
and the other being (preferred by @sto3psl )
```js
httpReq(path, options) // method is part of options here
```

Pretty sure this will cause some outrage and perhaps we'll settle on one of them instead of hacking `arguments` ;-)

Also added tests (against httpbin.org, which was easier than firing up our own echo server)

gka commented:
Maybe this is a good time to lay out my arguments for the `httpReq(method, path, options)` signature!

I know that many libraries including native fetch are doing this differently but please hear me out 😅

The thing I like most about this is that it forces the programmer to pass the full api route. `GET /v3/charts` and `POST /v3/charts` are radically different requests and they are handled by different handlers, whereas two `POST /v3/charts` requests with different payloads are basically the same requests. The three argument signature makes these differences clearer by forcing us to spell out the GET.

It also makes it clearer that we’re calling an *HTTP* API here I think. The HTTP verbs are not just some minor detail of the HTTP protocol but they are the front and center of it. It’s what every request has to begin with. I really like how this signature mirrors the raw http request!  Being able to omit the http method sort if diminishes that special status. 

Being a convenience fetch wrapper designed specifically for dealing with JSON APIs (hence the defaults for baseUrl and headers/cors/credentials) I also like how this signature highlights the REST design. 

Finally I think by passing the method as first argument we are rightfully removing the special status of GET requests over the other methods. I don’t really see that in our context GET requests are necessarily a more common use case than other verbs. Our app is likely using PATCH and PUT as often as GET. Let’s stop the discrimination of the other verbs! 😂

I think these are compelling arguments. Curious for your thoughts @sto3psl @ilokhov @sjockers @davidkokkelink

ilokhov commented:
I don't have a particularly strong preference for either way of passing arguments, but I do think it's important that we settle on just one of these for clarity.

I'm more drawn towards `httpReq(path, options)`, simply because it's the same as the native fetch API, however @gka also makes some valid arguments for explicitly including `method`.

sto3psl commented:
First of all, this is a great idea! I love that you already integrated parts of my lunch feedback! 👍 

Now let's talk about some parts of the public API 😄.  
As you know, I'm not that fond of the  `function (verb, path, options)` signature and mangling `arguments` is even worse 😅 . `verb` is in my opinion only an option. I was curious what other request libraries do and checked some:

|Library|`options.method`|method function (`library.method()`)|
|-|-|-|
`request`|✅|✅ `request.get()`|
`superagent`|❌|✅ `superagent.get()`|
`axios`|✅|✅ `axios.get()`|
`ky`|✅|✅ `ky.get()`|
`got`|✅|✅ `got.get()`|
`jquery`|✅ `jquery.ajax`|✅ `jquery.get()`|

As one can see, all of them use basically the same public API but instead of putting `verb` as a parameter, `verb`s have their own methods. I do like that and would love to hear your opinions on that.
Something else I realized is that with all of these libraries having a similar API, developers will already be familiar with `{ method: 'POST' }` or `library.post()`, when they ask how the wrapper works, it's just like all the others.

**In conclusion: I can see your desire to put `verb` first but would prefer it the way common libraries do it.**

---

About one of your assumptions of `GET`s special status:
> I don’t really see that in our context GET requests are necessarily a more common use case than other verbs. Our app is likely using PATCH and PUT as often as GET.

I fully agree that `PATCH` and `PUT` are probably called the most by our app but written in code, it might be different. There is no `PATCH` without `GET`ting the thing you want to patch. This PR is about writing code, to do network requests easier, and for every `PATCH` request, at some point there was a `GET` request written. The other way around is not necessarily the case.

---

In the end we should agree on one type, like @ivan said. I like what the mentioned libraries do and think it checks your requirements @gka.

gka commented:
I think i’d be happy with the verb functions and an option to pass the method as argument. 

so one could either call `httpReq(path, options)` or `httpReq.put(path, options)`. In case of the verb functions do you think we should ignore the `options.method` if set? or throw an error? don’t like the idea of overwriting the method in a `put()` call

sto3psl commented:
I tend to making it throw. If one uses `httpReq.put('/', { method: 'GET' })` that should be punished 🤪 and print a helpful error.

sto3psl commented on httpReq.js:
```diff
@@ -0,0 +1,81 @@
+
+function httpReqVerb(method) {
+    return (path, options) => {
+        if (options && options.method.toLowerCase() != method) {
```
```suggestion
        if (options && options.method) {
```

I would throw as soon as `options.method` is defined in these. It makes no sense and is redundant to do `httpReq.get('/', { method: 'get' })`

sto3psl commented on httpReq.js:
```diff
@@ -0,0 +1,81 @@
+            'Content-Type': 'application/json'
+        },
+        ...options
+    };
```
This is a fun one with object rest spread (and `Object.assign`). If I now call `httpReq` like this:
```js
httpReq('/', {
  headers: { 'Accept-Encoding': 'gzip' }
})
```
`options.headers` will overwrite the default `Content-Type` header. If you want this to be possible you need to change order around:

```diff
const { payload, baseUrl, raw, ...opts } = {
    payload: null,
    raw: false,
    method: 'get',
    baseUrl: `//${dw.backend.__api_domain}`,
    mode: 'cors',
    credentials: 'include',
+   ...options,
    headers: {
        'Content-Type': 'application/json',
+       ...options.headers
    }
-   ...options
};

gka replied:
👍

gka commented:
added the [readme](https://github.com/datawrapper/shared/blob/b8c77c776722250b0ff8db4f4285353fa906f07b/README.md#httpReq) entry for `httpReq` and adjusted the API according to the discussion here

gka commented:
I'm planning to merge this tomorrow so if anyone has any more feedback, please let me know :)

davidkokkelink commented:
i did not check the implementation, but i'm a big fan of the API design, looks very handy

gka commented:
I updated the examples in the top comment with the recent implementation. you can also [check out the readme](https://github.com/datawrapper/shared/blob/httpreq/README.md#httpReq)

ilokhov commented:
I'm happy with this too and glad we could settle on an approach that satisfies everyone.

sjockers commented:
I also like the API 👍 

Also, kudos for handling responses according to content types, I remember this being a pain in the old fetch implementation.

sjockers commented on httpReq.test.js:
```diff
@@ -0,0 +1,60 @@
+test.before(t => {
+    window.fetch = fetch;
+    global.dw = { backend: { __api_domain: 'api.datawrapper.local' } };
+});
```
You could consider stubbing fetch in order to speed up and further isolate the test (see [fetch.test.js](https://github.com/datawrapper/shared/blob/master/fetch.test.js) for an example of how to do that.

gka replied:
I did consider but I didn't like the fact that I have to mock the responses. If so I would want a real "echo" server, ideally one that behaves like httpbin. In theory we could fire up [httpbin](https://github.com/postmanlabs/httpbin) itself but it's written in Python and I didn't want to throw that into our local test setup. There are some similar echo-servers in node but I wasn't really impressed by any of them so far. Also considered writing one myself but decided it would be out of scope for this PR.

sjockers replied:
That makes perfect sense.
